x/sync/network_server.go

```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
 	}
 
 	// override limit if it is greater than maxKeyValuesLimit
-	limit := req.Limit
-	if limit > maxKeyValuesLimit {
-		limit = maxKeyValuesLimit
+	keyLimit := req.KeyLimit
+	if keyLimit > maxKeyValuesLimit {
+		keyLimit = maxKeyValuesLimit
 	}
+	bytesLimit := int(req.BytesLimit)
+	if bytesLimit > constants.DefaultMaxMessageSize {
+		bytesLimit = constants.DefaultMaxMessageSize
+	}
+
+	// attempt to get a proof within the bytes limit
+	for keyLimit > 0 {
+		changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(keyLimit))
+		if err != nil {
+			// handle expected errors so clients cannot cause servers to spam warning logs.
+			if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
+				s.log.Debug(
+					"dropping invalid change proof request",
+					zap.Stringer("nodeID", nodeID),
+					zap.Uint32("requestID", requestID),
+					zap.Stringer("req", req),
+					zap.Error(err),
+				)
+				return nil // dropping request
+			}
+			return err
+		}
+
+		proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
 
-	changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(limit))
-	if err != nil {
-		// handle expected errors so clients cannot cause servers to spam warning logs.
-		if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
-			s.log.Debug(
-				"dropping invalid change proof request",
+		// ensure that the new limit is always smaller
+		keyLimit = uint16((len(changeProof.KeyValues) + len(changeProof.DeletedKeys)) / 2)
+
+		// estimate the bytes of the start and end proof to ensure that everything will fit into the bytesLimit
+		bytesEstimate := getBytesEstimateOfProofNodes(changeProof.StartProof)
+
+		// just the start proof is too large, so a proof is impossible
+		if bytesEstimate > int(req.BytesLimit) {
+			// errors are fatal, so log for the moment
+			s.log.Warn(
+				"cannot generate a proof within bytes limit",
 				zap.Stringer("nodeID", nodeID),
 				zap.Uint32("requestID", requestID),
 				zap.Stringer("req", req),
-				zap.Error(err),
+				zap.Error(ErrMinProofSizeIsTooLarge),
 			)
-			return nil // dropping request
+			return nil
 		}
-		return err
-	}
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		bytesEstimate += getBytesEstimateOfProofNodes(changeProof.EndProof)
+		deleteKeyIndex := 0
+		changeKeyIndex := 0
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
```

Comment:
Instead of adding key values / deleted keys until the proof is too large, would it make more sense to _delete_ key values until the proof is small enough?

Reply:
it would require me to already know the total size, so I would have to add up all the values first.

Reply:
>We are eliminating half the key/values before we get here

The current heuristic is:

```
for {
    // Get proof
    // If proof small enough, return
    // Halve max number of keys
    // Check if proof will still be too big even with number of keys halved. If so, reduce number of keys more.
}
```

What I'm wondering is: what if instead of halving the number of keys, we just iterate over the proof we got, removing the greatest keys until the proof is small enough?

Reply:
Having said all that, we could just do the key shaving and try to add some level of buffer and just hope for the best.